### ISSUE TITLE
Add Row does not work with postgres identity columns that have many characters

### ISSUE THREAD

**Issue by Meibes** [qgis/QGIS#51563]
### What is the bug or the crash?

if you try to add a new row the value for the primary key is not pre-filled, sequences are ignored

### Steps to reproduce the issue

1. create example tables / database
```
CREATE DATABASE test_create_features;

CREATE TABLE test_identifier_name_so_long_it_will_get_truncated(
    test_identifier_so_long_it_will_get_truncated integer PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY,
    test_description text,
    test_default timestamp with time zone DEFAULT CURRENT_TIMESTAMP NOT NULL
);

CREATE TABLE test_serial_name_so_long_it_will_get_truncated(
    test_serial_so_long_it_will_get_truncated SERIAL PRIMARY KEY,
    test_description text,
    test_default timestamp with time zone DEFAULT CURRENT_TIMESTAMP NOT NULL
);

CREATE TABLE test_name(
    test_identifier integer PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY,
    test_description text,
    test_default timestamp with time zone DEFAULT CURRENT_TIMESTAMP NOT NULL
);
```

2. load the tables into QGIS
3. toggle editing for each table and click on Add Row

you can notice that the primary key column is prefilled for the "test_name" and serial table, but it does not work for the identity table with a lot of characters.

digging into the code I think I found the relevant part where the bug could be located: https://github.com/qgis/QGIS/blob/0a1f4332eb5b2c473d9e94e3fea803391f59cb50/src/providers/postgres/qgspostgresprovider.cpp#L1420-L1434

the seqName used to search for the sequence (mTableName + '_' + fieldName + QStringLiteral( "_seq" )) is just an educated guess used in the query. searching the internet I found a query that lists all sequences: https://dba.stackexchange.com/a/260977

```
SELECT t.oid::regclass AS table_name,
       a.attname AS column_name,
       s.relname AS sequence_name
FROM pg_class AS t
   JOIN pg_attribute AS a
      ON a.attrelid = t.oid
   JOIN pg_depend AS d
      ON d.refobjid = t.oid
         AND d.refobjsubid = a.attnum
   JOIN pg_class AS s
      ON s.oid = d.objid
WHERE d.classid = 'pg_catalog.pg_class'::regclass
  AND d.refclassid = 'pg_catalog.pg_class'::regclass
  AND d.deptype IN ('i', 'a')
  AND t.relkind IN ('r', 'P')
  AND s.relkind = 'S';
```

this query could be limited to the table in need by adding `AND t.oid::regclass = 'test_name'::regclass` for example, where 'test_name' is the name of one of my example tables. the sequence could then be retrieved by simply reading the result in 'sequence_name' for example and no educated guess would be necessary

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /></head><body>

QGIS-Version | 3.22.14-Białowieża | QGIS-Codeversion | 4cde646c
-- | -- | -- | --
Qt-Version | 5.15.3
Python-Version | 3.9.5
Kompiliert mit GDAL/OGR | 3.6.1 | Läuft mit GDAL/OGR | 3.6.2
PROJ-Version | 9.1.1
EPSG-Registraturdatenbankversion | v10.076 (2022-08-31)
GEOS-Version | 3.11.1-CAPI-1.17.1
SQLite-Version | 3.39.4
PDAL-Version | 2.4.3
PostgreSQL-Client-Version | 14.3
SpatiaLite-Version | 5.0.1
QWT-Version | 6.1.6
QScintilla2-Version | 2.13.1
BS-Version | Windows 10 Version 2009

</body></html>

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

**Comment by elpaso** [comment]
I agree it's a bug, it just doesn't seems so easy to fix as you might think, for instance what if the table contains more sequences? Wouldn't it be a guess in that case?

**Comment by Meibes** [comment]
> I agree it's a bug, it just doesn't seems so easy to fix as you might think, for instance what if the table contains more sequences? Wouldn't it be a guess in that case?

you could even further limit the results by adding `AND a.attname = 'test_identifier'` to only get results for the table and column in question.

**_better idea:_**
while thinking about this issue I also stumbled upon a probably much cleaner approach, using an already existing function: [pg_get_serial_sequence ( table text, column text ) → text](https://www.postgresql.org/docs/current/functions-info.html#FUNCTIONS-INFO-CATALOG-TABLE)

> Returns the name of the sequence associated with a column, or NULL if no sequence is associated with the column. If the column is an identity column, the associated sequence is the sequence internally created for that column. For columns created using one of the serial types (serial, smallserial, bigserial), it is the sequence created for that serial column definition. [...] The first parameter is a table name with optional schema, and the second parameter is a column name. Because the first parameter potentially contains both schema and table names, it is parsed per usual SQL rules, meaning it is lower-cased by default. The second parameter, being just a column name, is treated literally and so has its case preserved. [...]

with my example data, the queries would look like: 
```
SELECT pg_get_serial_sequence('public.test_identifier_name_so_long_it_will_get_truncated', 'test_identifier_so_long_it_will_get_truncated');
SELECT pg_get_serial_sequence('public.test_serial_name_so_long_it_will_get_truncated', 'test_serial_so_long_it_will_get_truncated');
SELECT pg_get_serial_sequence('public.test_name', 'test_identifier');
```

this query does always return only one sequence name and it favors the sequence that was created by a serial or the identity, not the ones later added by the user (in case there are several sequences on a single column, which the current solution simply ignores and does not return anything: `if ( seqResult.PQntuples() == 1 )`).

this query is easier to understand and to maintain than the current approach in my opinion

**Comment by elpaso** [comment]
I'll have a look, thank you for the suggestions!

**Comment by Meibes** [comment]
@elpaso thanks for the quick fix! :)

**Comment by elpaso** [comment]
@Meibes thank you for the diagnosis and the solution! I only had to implement (and test) it.

**Comment by Meibes** [comment]
just a quick question: the qgis-bot closed the pull request #51649, does this mean this fix will not get into 3.28.4?

**Comment by elpaso** [comment]
> just a quick question: the qgis-bot closed the pull request #51649, does this mean this fix will not get into 3.28.4?

Yes,  unless some other developer approves it. I can reopen the PR but I shouldn't self-approve it.
